### PR TITLE
Resetting managed fields and fieldtype

### LINKED ISSUE #90610: Updates with client-go dropping ManagedFieldsEntry.FieldsType breaks
https://github.com/kubedb/issues/issues/739

[`ManagedFieldsEntry. FieldsType`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1123) is not defaulted and can be dropped by pre 1.15 versions of client-go (since the field didn't exist at that time). Unfortunately, sending an update will then result in a validation failure since the field is required.

We should figure out what to do when we receive a `ManagedFieldsEntry` whose `FieldsType` is not set.

/sig api-machinery
/priority important-soon

### PR DESCRIPTION
**Pull request by apelisse**
Addresses two of the important issues:
- Changes the way we reset fields either by sending empty list or, to keep compatibility, send an empty item.
- If FieldsType is empty, do the right thing (which is, fallback to liveObject) and fallback to empty if that doesn't work either.

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #90610

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Resolve regression in metadata.managedFields handling in update/patch requests submitted by older API clients
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go [hunk 1]
```diff
@@ -138,17 +144,32 @@ func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (o
 	return object, nil
 }
 
+// Returns true if the managedFields that have indicate that the user is trying to reset the managedFields.
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:147]: I'll fix that :-)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go [hunk 2]
```diff
@@ -138,17 +144,34 @@ func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (o
...
+// Returns true if the managedFields indicate that the user is trying to
+// reset the managedFields, i.e. if the list is non-nil but empty, or if
+// the list has one empty item.
+func isResetManagedFields(managedFields []metav1.ManagedFieldsEntry) bool {
+	if len(managedFields) == 0 {
+		return managedFields != nil
+	}
+
+	if len(managedFields) == 1 {
+		mf := managedFields[0]
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:156]: why not `reflect.DeepEqual(managedFields[0], metav1.ManagedFieldsEntry{})`?

**Comment by liggitt** [comment]
one comment about the empty item check, then lgtm

/hold for https://github.com/kubernetes/kubernetes/pull/91690

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:156]: Done

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:156]: did this get pushed?

Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go]: I must have been interrupted somehow? Sorry about that ...

**Comment by apelisse** [comment]
I think hold can be removed since the other PR was merged.

thanks, PTAL @liggitt

**Comment by apelisse** [comment]
OK I'll cherry-pick this in 1.18, 1.17 and 1.16.

**Pull request merged**